### ISSUE TITLE
[feature] Add support for appending a provenance attestation to an existing one

### ISSUE THREAD

**Issue by sethmlarson** [slsa-framework/slsa-github-generator#1565]
**Is your feature request related to a problem? Please describe.**

Python is different in that binary wheels sometimes require building artifacts "per Python version" (ie `MarkupSafe-2.1.2-cp311-...whl` is only for Python 3.11). This means that a package that wants to support new Python versions will need to build new wheels but don't necessarily want to make a brand new release because the source code hasn't changed. This saves users from needlessly incrementing versions/lockfiles.

In order to support this use-case, provenance files should be appendable with new attestations after the initial release and attestation is published. Since the provenance file is `.jsonl` I think they can be appended directly to the file.

The `MarkupSafe` project suggested implementing this via the `workflow_dispatch` GitHub Action trigger to allow updating the `cibuil

**Describe the solution you'd like**

- Adding functionality for appending the new provenance attestation to an existing file.
- Adding an option for performing the above operation to a GitHub release with an attached provenance attestation.

**Describe alternatives you've considered**

- Making a brand new release even if source code hasn't changed.
- Rebuilding all artifacts and replacing the provenance attestation in-place for the GitHub Release. This assumes that your build is reproducible.

**Additional context**

Related: https://github.com/slsa-framework/slsa-verifier/issues/454

**Comment by laurentsimon** [comment]
Thanks @sethmlarson 
From a user's perspective, will they need to have 2 workflows to release? One that works on new tags, and one that is called manually with a tag (workflow_dispatch input)?

**Comment by sethmlarson** [comment]
@laurentsimon It could be one workflow that triggers on either tag or workflow_dispatch, that's how I initially imagined things since duplicate wheels aren't a problem and won't get published to Gh/PyPI.

If it could be different flows without a problem then that would be handy too!

**Comment by laurentsimon** [comment]
Thanks. The reason I ask is for consistency of the tag release. In a workflow dispatch, it's manually provided; in a new tag trigger, it's provided by the GH context. This means verification could be different for the user.

Note for the implementation: appending to an existing provenance file incurs a risk of corruption / race condition if multiple workflows are running concurrently

**Comment by ianlewis** [comment]
> Note for the implementation: appending to an existing provenance file incurs a risk of corruption / race condition if multiple workflows are running concurrently

Note that we kind of have this currently. i.e. if two runs of the same workflow at the same tag are running concurrently
1. Workflow run 1 uploads artifact to a release
2. Workflow run 2 uploads (and overwrites) artifact
3. Workflow run 2 uploads attestation (jsonl) the release
4. Workflow run 1 uploads (and overwrites) attestation

In this case, verification will fail if the artifacts aren't byte-for-byte the same (as it should).

So I don't think it's a huge deal different from what we have currently.

**Comment by laurentsimon** [comment]
@sethmlarson did you need to append to an existing attestation or create a new attestation for an existing release?

If the former, the race condition is a problem. 
If the latter, I think what @ianlewis works and you'd be in charge of creating an attestation with a name that does not collide with existing assets.

From the description, it looks like you would like both?

**Comment by ianlewis** [comment]
Right, I'm assuming what happens is a new release asset is created and a new provenance and attestation are created and we want to add that to an existing release. In that case we can add it to the existing jsonl file somehow or just add a new jsonl (with a name that makes it obvious what jsonl file pertains to which whl).

**Comment by sethmlarson** [comment]
The original intent was to add the new attestation to an existing release with an existing .jsonl file so if the documentation describes a single `multiple.intoto.jsonl` to be used for verification users can continue downloading that same file.

**Comment by davidism** [comment]
SLSA 1.0 now says that one provenance file should be generated per artifact, like `filename.intoto.jsonl` per filename, rather than `multiple.intoto.jsonl`. Maybe instead of appending we need a `provenance-per-filename` setting instead. I can open an issue about that if it makes sense.

**Comment by davidism** [comment]
I've opened https://github.com/slsa-framework/slsa-github-generator/issues/2116.
